// scandit-sdk/rollup.config.js
// rollup.min-config.js
import autoprefixer from "autoprefixer";
import babel from "rollup-plugin-babel";
import commonjs from "rollup-plugin-commonjs";
import cssnano from "cssnano";
import nodeResolve from "rollup-plugin-node-resolve";
import nodeSass from "node-sass";
import postcss from "postcss";
import sass from "rollup-plugin-sass";
import sourcemaps from "rollup-plugin-sourcemaps";
import typescript from "rollup-plugin-typescript2";
import { terser } from "rollup-plugin-terser";

import { version } from "./package.json";

const config = {
  input: "src/index.ts",
  output: {
    sourcemap: true,
    banner: `/*!
 * @license
 *
 * Scandit Barcode Scanner SDK for the Web
 * v. ${version}
 *
 * Copyright © 2019 Scandit AG. All Rights Reserved.
 *
 * The use of this software is governed by the Scandit Terms and Conditions.
 * https://ssl.scandit.com/terms/test.pdf
 *
 * The following sets forth attribution notices for third party software that may be contained in portions of the product.
 * https://docs.scandit.com/stable/web/LICENSE
 */`
  },
  plugins: [
    nodeResolve({
      mainFields: ["module", "main", "browser"]
    }),
    commonjs({
      namedExports: {
        "node_modules/eventemitter3/index.js": ["EventEmitter"]
      }
    }),
    sass({
      insert: true,
      runtime: nodeSass,
      processor: css => {
        return postcss([autoprefixer, cssnano])
          .process(css, {
            from: undefined
          })
          .then(result => {
            return result.css;
          });
      }
    }),
    typescript({
      tsconfig: "./tsconfig.module.json",
      typescript: require("typescript"),
      cacheRoot: "./node_modules/.cache/.rts2_cache/" + process.env.CACHE_FOLDER,
      useTsconfigDeclarationDir: true
    }),
    babel({
      extensions: [".js", ".jsx", "ts", "tsx"],
      presets: [
        [
          "@babel/preset-env",
          {
            targets: ["last 3 versions", "> 0.1%", "not IE < 9"],
            modules: false,
            useBuiltIns: "usage",
            corejs: 2
          }
        ]
      ],
      plugins: [
        [
          "@babel/plugin-transform-runtime",
          {
            corejs: 2
          }
        ]
      ],
      overrides: [
        {
          test: "**/engineWorker.ts",
          presets: [
            [
              "@babel/preset-env",
              {
                useBuiltIns: false
              }
            ]
          ],
          plugins: [
            [
              "@babel/plugin-transform-runtime",
              {
                corejs: false,
                helpers: false,
                regenerator: false
              }
            ]
          ]
        }
      ],
      babelrc: false,
      runtimeHelpers: true,
      exclude: "node_modules/**"
    })
  ],
  onwarn: function(message) {
    if (/Circular dependency:/.test(message)) {
      return;
    }
    console.error(message);
  }
};

if (process.env.NODE_ENV === "production") {
  config.plugins.push(
    terser({
      output: {
        comments: /Scandit Barcode Scanner SDK for the Web/i
      },
      compress: {
        passes: 3
      }
    })
  );
  config.plugins.push(sourcemaps());
} else {
  config.plugins.push(sourcemaps());
}

export default config;
